### PR TITLE
Fixing performance issues in current BWMF task implementation.

### PR DESCRIPTION
**Pull request by BaiGang**
This PR includes modifications and improvements:

* A unified sparse representation of matrix shard. We do not need dense representation any more, instead we use a vec parameter to store param data and copy the values to matrix shard with 0s filtered.
* An improved implementation of KL divergence loss. We now do one pass to evaluate the loss and accumulate the gradient.
* T/D shard are initialized as sparse/random matrices if not loaded.

### CODE REVIEW

**Comment by BaiGang** [comment]
Now the cost of evaluation KL-DIv<V, W*H> is O(m*n*nonzero(W)), where V is mxn, W is mxk and H is kxn.

File: example/bwmf/bwmf_task.go [hunk 1]
```diff
@@ -113,37 +115,84 @@ func (t *bwmfTask) initData() {
...
+		t.dShard = &pb.MatrixShard{
+			Row: make([]*pb.MatrixShard_RowData, t.dims.m),
+		}
+		m2 := int32((10 + t.dims.m) / 10)
+		for r, _ := range t.dShard.Row {
+			t.dShard.Row[r] = &pb.MatrixShard_RowData{At: make(map[int32]float32)}
+
+			if rand.Int31()%m2 == 0 {
+				for k := 0; k < t.dims.k; k++ {
+					t.tShard.Row[r].At[int32(k)] = rand.Float32()
```
Comment (5kg) [inline example/bwmf/bwmf_task.go]: s/tShard/dShard/

**Comment by 5kg** [comment]
> Now the cost of evaluation KL-DIv is O(mnnonzero(W)), where V is mxn, W is mxk and H is kxn.

I saw an improvement 122.460s -> 95.141s

**Comment by 5kg** [comment]
I tried to understand the effects of these series of changes.

* with 35054ee59696e27827323e67c84cbb87735f5df0 alone, I saw `457.892s -> 119.28s`
* with the spareness patch 64f4fa3a503fcebf0b536db04b0f95dd481c2764 (no thresholding, no new initialization method): `119.28s -> 150.105s`
* with thresholding: `150.105s -> 134.245s`
* with new initialization method: `134.245s -> 76.420s`, BUT performance get worse (in term of loss `80879.351562 v.s. 48503.832031`)

**Comment by 5kg** [comment]
New profiling result, profile

map access is now the hot-spot.

**Comment by BaiGang** [comment]
We got a build error in [the trivis build run](https://travis-ci.org/taskgraph/taskgraph/builds/60640479) for commit 0e111ca:

```
../../colinmarc/hdfs/client.go:4:2: cannot find package "code.google.com/p/goprotobuf/proto" in any of:
	/home/travis/.gimme/versions/go1.4.linux.amd64/src/code.google.com/p/goprotobuf/proto (from $GOROOT)
	/home/travis/gopath/src/code.google.com/p/goprotobuf/proto (from $GOPATH)
The command ".script/test" exited with 1.
```

It seems irrelevant here.

**Comment by xiaoyunwu** [comment]
@hongchao, can you take a look?

Xiaoyun

**Comment by fengjingchao** [comment]
No. We don't have that dependency.
Just add any dependency we want to ".script/dep.sh"

**Comment by BaiGang** [comment]
Next, we will be working on a parallel version of `KLDivLoss.Evaluate()`. @5kg

**Comment by fengjingchao** [comment]
Is this ready to commit? And can you remove the binary files?

**Comment by BaiGang** [comment]
I think it is ready to merge. @xiaoyunwu Please take a look if it's in the
right direction. Next we'll be parallelizing the evaluation and
accumulation of loss and gradients in kldiv_loss.go.

As for the binary files, they are actually the marshalled data of matrix
shard for integration test. Shall we generate these files on the fly during
integration test, or just leave it as is?

**Comment by fengjingchao** [comment]
Keep it if you need.
I remove some test generated files in script:
https://github.com/taskgraph/taskgraph/blob/master/.script/test#L10

I think those files should be generated in a dir like "tmp/" to be separated from "data/", and add the dir to gitignore.

**Comment by BaiGang** [comment]
OK. I'll modify it to generate test data in `tmp/` and push it later.

**Comment by fengjingchao** [comment]
LGTM.

Thanks everyone!

**Pull request merged**